core/src/main/java/net/opentsdb/query/processor/downsample/DownsampleNumericToNumericArrayIterator.java

```diff
@@ -0,0 +1,577 @@
+      double[] localDoubleAggs) {
+    if (!accumulator.hasLongValue) {
+      return;
+    }
+    // run accumulator for longs
+    for (int i = 0; i < accumulator.longIndex; i++) {
+
+      long v = accumulator.longValues[i];
+
+      switch (this.aggEnum) {
+        case sum:
+          localLongAggs[0] += v;
+          localLongAggs[5] = 0; // dirty flag
+          break;
+        case count:
+          localLongAggs[1]++;
+          localLongAggs[5] = 0; // dirty flag
+          break;
+        case min:
+          if (v < localLongAggs[2]) {
+            localLongAggs[2] = v;
+          }
+          localLongAggs[5] = 0; // dirty flag
+          break;
+        case max:
+          if (v > localLongAggs[3]) {
+            localLongAggs[3] = v;
+          }
+          localLongAggs[5] = 0; // dirty flag
+          break;
+        case last:
+          localLongAggs[4] = v;
+          localLongAggs[5] = 0; // dirty flag
+          localDoubleAggs[5] = -1; // Reset double dirty flag.
+          break;
+        case avg:
+          localLongAggs[0] += v;
+          localLongAggs[1]++;
+          localLongAggs[5] = 0; // dirty flag
+          break;
+
+        default:
+          throw new UnsupportedOperationException(
+              "Unsupported aggregator function: " + this.aggEnum.name());
+      }
+    }
+  }
+
+  private void accumulateDoubles(Accumulator accumulator, long[] localLongAggs,
+      double[] localDoubleAggs) {
+    if (!accumulator.hasDoubleValue) {
+      return;
+    }
+    // run accumulator for doubles
+    for (int i = 0; i < accumulator.doubleIndex; i++) {
+
+      double v = accumulator.doubleValues[i];
+
+      // Ignore NaNs in dowmsampling
+      if (Double.isNaN(v)) {
+        continue;
+      }
+
+      switch (this.aggEnum) {
+        case sum:
+          localDoubleAggs[0] += v;
+          localDoubleAggs[5] = 0; // dirty flag
+          break;
+        case count:
+          localDoubleAggs[1]++;
+          localDoubleAggs[5] = 0; // dirty flag
+
+          break;
+        case min:
+          if (v < localDoubleAggs[2]) {
+            localDoubleAggs[2] = v;
+          }
+          localDoubleAggs[5] = 0; // dirty flag
+          break;
+        case max:
+          if (v > localDoubleAggs[3]) {
+            localDoubleAggs[3] = v;
+          }
+          localDoubleAggs[5] = 0; // dirty flag
+          break;
+        case last:
+          localDoubleAggs[4] = v;
+          localDoubleAggs[5] = 0; // dirty flag
+          localLongAggs[5] = -1; // Reset long agg
+          break;
+        case avg:
+          localDoubleAggs[0] += v;
+          localDoubleAggs[1]++;
+          localDoubleAggs[5] = 0; // dirty flag
+          break;
+
+        default:
+          throw new UnsupportedOperationException(
+              "Unsupported aggregator function: " + this.aggEnum.name());
+      }
+    }
+  }
+
+  static class Accumulator {
+    private final double[] doubleValues;
```

Comment:
To genericize it we should move this into the aggregators classes and use those interfaces.